### PR TITLE
Fix cycle detection behavior in Analyzer

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by sjrd** [comment]
What's the issue that is fixed by the first commit? Can we write a test for it in [`AnalyzerTest`](https://github.com/scala-js/scala-js/blob/master/linker/shared/src/test/scala/org/scalajs/linker/AnalyzerTest.scala)?

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala [hunk 1]
```diff
@@ -175,7 +175,7 @@ object Analysis {
 
   final case class MissingJavaLangObjectClass(from: From) extends Error
   final case class InvalidJavaLangObjectClass(from: From) extends Error
-  final case class CycleInInheritanceChain(cycle: List[ClassInfo], from: From) extends Error
+  final case class CycleInInheritanceChain(cycle: List[Infos.ClassInfo], from: From) extends Error
```
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: The problem is not the amount of information but the typing thereof. `ClassInfo#superClass` returns a `Option[ClassInfo]`. `Infos.ClassInfo#superClass` returns a `Option[String]`.

[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: We could entirely hide `superClass`, though.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala [hunk 2]
```diff
@@ -283,12 +283,18 @@ private final class Analyzer(config: CommonPhaseConfig,
...
       assert(!linked)
       linkedFrom ::= from
+
+      val Self = this
       try {
         linkClassesImpl()
       } catch {
-        case CyclicDependencyException(chain, _) =>
-          throw CyclicDependencyException(this :: chain, from)
+        case CyclicDependencyException(cycle, Self, null) =>
```
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala]: Done

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: IMO I would mostly care about the `cycle` list itself. There's not much handling to do except report the cycle to the user anyway. The user will figure out easily enough whether a parent in the parent chain is a superclass or an interface. So yes I believe the name should be enough.

[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: Done.

**Comment by gzm0** [comment]
Updated

**Pull request merged**